cf_xarray/accessor.py

```diff
@@ -1523,6 +1524,112 @@ def differentiate(
+        ----------
+        override: bool
+            Override existing attributes
+        skip: str, iterable, optional
+            Attribute keys to skip: {"units", "grib", "amip", "description"}
+        verbose: bool
+            Print added attributes to screen
+        source: optional
+            Path of cf-standard-name-table.xml or file object containing XML data.
+            If None, use the default CF standard name table.
+
+        Returns
+        -------
+        DataArray or Dataset with attributes added
+        """
+
+        # Defaults
+        skip = skip or []
+        root = (
+            ElementTree.parse(source).getroot()
+            if source
+            else ElementTree.fromstring(CF_STANDARD_NAME_TABLE)
+        )
+
+        # Construct table
+        info = {}
+        table: dict = {}
+        aliases = {}
+        for child in root:
+            if child.tag == "entry":
+                key = child.attrib.get("id")
+                table[key] = {}
+                for item in ["canonical_units", "grib", "amip", "description"]:
+                    parsed = child.findall(item)
+                    attr = item.replace("canonical_", "")
+                    table[key][attr] = (parsed[0].text or "") if parsed else ""
+            elif child.tag == "alias":
+                alias = child.attrib.get("id")
+                key = child.findall("entry_id")[0].text
+                aliases[alias] = key
+            else:
+                info[child.tag] = child.text
+
+        # Loop over standard names
+        ds = self._maybe_to_dataset().copy()
+        attrs_to_print: dict = {}
+        for std_name, var_names in ds.cf.standard_names.items():
+
+            # Loop over variable names
+            for var_name in var_names:
+                old_attrs = ds[var_name].attrs
+                std_name = aliases.get(std_name, std_name)
+                new_attrs = table.get(std_name, {})
+
+                # Loop over attributes
+                for key, value in new_attrs.items():
+                    if value and key not in skip and (override or key not in old_attrs):
+
+                        # Don't add units to time variables (e.g., datetime64, ...)
+                        if key == "units" and _is_datetime_like(ds[var_name]):
+                            continue
+
+                        # Add attribute
+                        ds[var_name].attrs[key] = value
+
+                        # Build verbose dictionary
+                        attrs_to_print.setdefault(var_name, {})
+                        attrs_to_print[var_name][key] = value
+
+                        # Update history
+                        history = ds[var_name].attrs.get("cf_history", "")
+                        if history:
+                            history += "; "
+                        history += f"add canonical CF attribute {key!r} (v{info['version_number']!s})"
+                        ds[var_name].attrs["cf_history"] = history
```

Comment:
Thoughts?
It's kind of quick and dirty, I guess we will eventually have a more robust approach for all methods?
What do you think about using a separate `cf_history` rather than `history`?

Reply:
`history` is an NCO convention I think. We'll want to stick this new statement at the front of a history attribute, if any

Reply:
If we add to `history` we should probably align with existing conventions:
1. I think NCO only adds global attributes, should we do that with Datasets?
2. To align with NCO I guess we'd have to add `date: .cf.add_canonical_cf_attributes(arg1=..., arg2=...)` and a separate attribute with cf-xarray and cf-table version?
```
:history = "Wed Feb 24 15:32:09 2021: ncks --no_abc --cnk_dmn x,64 --cnk_dmn y,64 --cnk_dmn nav_lev,1 --cnk_dmn time_counter,1 --dfl_lvl 1 GSRIDGE36Z_mesh_mask.nc GSRIDGE36Z_mesh_mask.nc" ;
:NCO = "netCDF Operators version 4.7.6 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)" ;
```